server/src/main/java/org/elasticsearch/index/engine/ThreadPoolMergeExecutorService.java

```diff
@@ -74,27 +186,57 @@ public class ThreadPoolMergeExecutorService {
     private final int maxConcurrentMerges;
     private final int concurrentMergesFloorLimitForThrottling;
     private final int concurrentMergesCeilLimitForThrottling;
+    private final BudgetTracker<MergeTask> budgetTracker;
+    private final AvailableDiskSpacePeriodicMonitor availableDiskSpacePeriodicMonitor;
 
     private final List<MergeEventListener> mergeEventListeners = new CopyOnWriteArrayList<>();
 
     public static @Nullable ThreadPoolMergeExecutorService maybeCreateThreadPoolMergeExecutorService(
         ThreadPool threadPool,
-        Settings settings
+        Settings settings,
+        ClusterSettings clusterSettings,
+        NodeEnvironment nodeEnvironment
     ) {
         if (ThreadPoolMergeScheduler.USE_THREAD_POOL_MERGE_SCHEDULER_SETTING.get(settings)) {
-            return new ThreadPoolMergeExecutorService(threadPool);
+            return new ThreadPoolMergeExecutorService(threadPool, settings, clusterSettings, nodeEnvironment);
         } else {
             return null;
         }
     }
 
-    private ThreadPoolMergeExecutorService(ThreadPool threadPool) {
+    private ThreadPoolMergeExecutorService(
+        ThreadPool threadPool,
+        Settings settings,
+        ClusterSettings clusterSettings,
+        NodeEnvironment nodeEnvironment
+    ) {
         this.executorService = threadPool.executor(ThreadPool.Names.MERGE);
         this.maxConcurrentMerges = threadPool.info(ThreadPool.Names.MERGE).getMax();
         // the intent here is to throttle down whenever we submit a task and no other task is running
         this.concurrentMergesFloorLimitForThrottling = 2;
         this.concurrentMergesCeilLimitForThrottling = maxConcurrentMerges * 2;
         assert concurrentMergesFloorLimitForThrottling <= concurrentMergesCeilLimitForThrottling;
+        this.budgetTracker = new BudgetTracker<>(MergeTask::estimatedRemainingMergeSize, Long.MAX_VALUE, queuedMergeTasks::add);
```

Comment:
the last argument, i.e. `queuedMergeTasks::add`, is invoked for the over-budget merge tasks (as soon as more budget becomes available).